src/pulselistener/pulselistener/mercurial.py

```diff
@@ -19,93 +20,263 @@
+
+        # Build asyncio shared queue
+        self.queue = asyncio.Queue()
+
+        # Configure repositories, and index them by phid
+        self.repositories = {
+            phab_repo['phid']: Repository(conf, cache_root)
+            for phab_repo in self.phabricator_api.list_repositories()
+            for conf in repositories
+            if phab_repo['fields']['name'] == conf['name']
+        }
+        assert len(self.repositories) > 0, 'No repositories configured'
+        logger.info('Configured repositories', names=[r.name for r in self.repositories.values()])
+
+    async def run(self):
+        # First clone all repositories
+        for repo in self.repositories.values():
+            logger.info('Cloning repo {}'.format(repo))
+            await repo.clone()
+
         # Wait for phabricator diffs to apply
         while True:
-            build_target_phid, diff = await self.queue.get()
-            assert isinstance(diff, dict)
-            assert 'phid' in diff
+            build = await self.queue.get()
+            assert isinstance(build, PhabricatorBuild)
+
+            # Find the repository from the diff and trigger the build on it
+            repository = self.repositories.get(build.repo_phid)
+            if repository is not None:
+                await self.handle_build(repository, build)
 
-            await self.handle_build(build_target_phid, diff)
+            else:
+                logger.error('Unsupported repository', repo=build.repo_phid, build=build)
 
             # Notify the queue that the message has been processed
             self.queue.task_done()
 
-    async def handle_build(self, build_target_phid, diff):
+    async def handle_build(self, repository, build):
         '''
         Try to load and apply a diff on local clone
         If succesful, push to try and send a treeherder link
         If failure, send a unit result with a warning message
         '''
+        assert isinstance(repository, Repository)
         failure = None
         start = time.time()
+
         try:
-            await self.push_to_try(build_target_phid, diff)
+            # Start by cleaning the repo
+            repository.clean()
+
+            # Get the stack of patches
+            base, patches = self.phabricator_api.load_patches_stack(repository.repo, build.diff, default_revision='central')
+            assert len(patches) > 0, 'No patches to apply'
+
+            # Load all the diffs details with commits messages
+            diffs = self.phabricator_api.search_diffs(
+                diff_phid=[p[0] for p in patches],
+                attachments={
+                    'commits': True,
+                }
+            )
+            commits = {
+                diff['phid']: diff['attachments']['commits'].get('commits', [])
+                for diff in diffs
+            }
+
+            await asyncio.sleep(2)  # allow other tasks to run
```

Comment:
We already have a wait in apply_patches, do we need another one here?

BTW, if the only reason for sleeping is allowing other tasks to run, we could just sleep for 0 seconds. This will allow other tasks to run without actually blocking the current task for too long.